### PR TITLE
feat(gen8): Wave 4 -- weather effects + entry hazards (G-Max Steelsurge, Heavy-Duty Boots)

### PR DESCRIPTION
**Pull request by uehlbran**
## Summary

- **Gen8Weather.ts**: Sandstorm/Hail chip damage (1/16 max HP), type/ability/item immunities (Ice Face new in Gen 8), weather duration (5/8 turns)
- **Gen8EntryHazards.ts**: Stealth Rock, Spikes, Toxic Spikes, Sticky Web, **G-Max Steelsurge** (Steel-type SR, new in Gen 8), **Heavy-Duty Boots** (blocks all hazards, new in Gen 8), Magic Guard interaction
- **Core type update**: Added `"gmax-steelsurge"` to `EntryHazardType` union
- **Gen8Ruleset.ts**: Wired `applyWeatherEffects` and `applyEntryHazards` overrides, added G-Max Steelsurge to available hazards
- 68 new tests (31 weather + 37 entry hazards), all passing

## Test plan

- [x] Weather: sandstorm/hail chip damage calculations (floor(maxHP/16))
- [x] Weather: type-based immunity (Rock/Ground/Steel for sand, Ice for hail)
- [x] Weather: ability-based immunity (Magic Guard, Overcoat, Ice Face, Slush Rush, etc.)
- [x] Weather: item-based immunity (Safety Goggles)
- [x] Weather: rain/sun produce no chip damage
- [x] Entry hazards: Stealth Rock at 5 effectiveness levels (0.25x through 4x)
- [x] Entry hazards: Spikes at 3 layer counts (1/8, 1/6, 1/4 max HP)
- [x] Entry hazards: Toxic Spikes (poison at 1 layer, toxic at 2, absorbed by grounded Poison)
- [x] Entry hazards: Sticky Web (-1 Speed, Defiant/Competitive interaction, Clear Body blocks)
- [x] Entry hazards: G-Max Steelsurge at 6 effectiveness levels
- [x] Entry hazards: Heavy-Duty Boots blocks all 5 hazard types
- [x] Entry hazards: Magic Guard blocks damage hazards but not Sticky Web
- [x] Entry hazards: Combined hazard application
- [x] All 276 gen8 tests passing
- [x] TypeScript typecheck passing
- [x] Biome lint passing

Closes: N/A

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Gen 8 entry hazards added, including G‑Max Steelsurge, full Heavy‑Duty Boots blocking, and Magic Guard interactions
  * Gen 8 weather end‑of‑turn chip damage for sandstorm and hail with type/ability/item immunities and configurable durations
  * Entry hazard ordering, stacking, Toxic Spikes absorption, and Sticky Web speed/stat interactions implemented

* **Tests**
  * Comprehensive test suites validating Gen 8 entry hazards, weather behavior, immunities, and edge cases

### CODE REVIEW

File: packages/gen8/tests/entry-hazards.test.ts [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+      // Source: Showdown data/moves.ts line 7475 -- gmaxsteelsurge: Steel-type Stealth Rock
+      // Steel vs Fire = 1x (Steel is not super effective against Fire, and Fire resists Steel -- 0.5x)
+      // Actually: Steel is NOT effective vs Fire. Steel vs Fire = 0.5x
+      // floor(200 * 0.5 / 8) = floor(12.5) = 12
+      const mon = makeActivePokemon({ types: ["fire"], maxHp: 200 });
+      const result = applyGen8GMaxSteelsurge(mon, GEN8_TYPE_CHART);
+      expect(result).not.toBeNull();
+      // Steel vs Fire = 0.5x; floor(200 * 0.5 / 8) = 12
+      expect(result!.damage).toBe(12);
+    });
```
[the remark this answers is not in the document]
Comment (uehlbran) [reply packages/gen8/tests/entry-hazards.test.ts]: Good catch. Fixed in b7e282a -- the test name now correctly says "Steel resisted by Fire = 0.5x" and "deals 12 damage" to match the assertion.

File: packages/gen8/tests/entry-hazards.test.ts [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+    it("given Sticky Web and a Pokemon with Clear Body, when applying, then blocked", () => {
+      // Source: Showdown data/abilities.ts -- clearbody: onBoost blocks stat drops
+      const mon = makeActivePokemon({ types: ["normal"], maxHp: 200, ability: "clear-body" });
+      const result = applyGen8StickyWeb(mon, false);
+      expect(result.applied).toBe(false);
+      expect(result.messages).toEqual(
+        expect.arrayContaining([expect.stringContaining("Clear Body prevents stat loss")]),
+      );
+    });
+  });
```
[the remark this answers is not in the document]
Comment (uehlbran) [reply packages/gen8/tests/entry-hazards.test.ts]: Good suggestion. Added a Competitive test in 3d19d0d verifying that -1 Speed triggers +2 Sp. Atk when Competitive is active, triangulating the stat-boost behavior alongside the existing Defiant test.

[the remark this answers is not in the document]
Comment (uehlbran) [reply packages/gen8/tests/entry-hazards.test.ts]: Noted. Competitive mirrors Defiant's logic (same code path, different stat). Will add a Competitive test case in a follow-up to improve triangulation.
